src/compression/b44/mod.rs

```diff
@@ -0,0 +1,924 @@
+        s[0] &= 0x7fff;
+    } else {
+        s[0] = !s[0];
+    }
+
+    for i in 0..16 {
+        s[i] = s[0];
+    }
+}
+
+#[derive(Debug)]
+struct ChannelData {
+    tmp_start_index: usize,
+    tmp_end_index: usize,
+
+    resolution: Vec2<usize>,
+    y_sampling: usize,
+    type_: SampleType,
+    quantize_linearly: bool,
+    samples_per_pixel: usize,
+}
+
+fn cpy_u16(src: &[u16], src_i: usize, dst: &mut [u16], dst_i: usize, n: usize) {
+    // assert_eq!(src.len(), dst.len());
+
+    // for i in 0..src.len() {
+    for i in 0..n {
+        dst[dst_i + i] = src[src_i + i];
+    }
+}
+
+pub fn decompress(
+    channels: &ChannelList,
+    compressed: &ByteVec,
+    rectangle: IntegerBounds,
+    expected_byte_size: usize,
+    _pedantic: bool,
+) -> Result<ByteVec> {
+    let expected_value_count = expected_byte_size / 2;
+    debug_assert_eq!(
+        expected_byte_size,
+        rectangle.size.area() * channels.bytes_per_pixel
+    );
+    debug_assert_ne!(expected_value_count, 0);
+    debug_assert!(!channels.list.is_empty());
+
+    if compressed.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    let mut channel_data: Vec<ChannelData> = Vec::with_capacity(channels.list.len());
+    let mut tmp_read_index = 0;
+
+    for channel in channels.list.iter() {
+        let channel = ChannelData {
+            tmp_start_index: tmp_read_index,
+            tmp_end_index: tmp_read_index,
+            resolution: channel.subsampled_resolution(rectangle.size),
+            y_sampling: channel.sampling.y(),
+            type_: channel.sample_type,
+            quantize_linearly: channel.quantize_linearly,
+            samples_per_pixel: channel.sample_type.bytes_per_sample()
+                / SampleType::F16.bytes_per_sample(),
+        };
+
+        tmp_read_index += channel.resolution.area() * channel.samples_per_pixel;
+        channel_data.push(channel);
+    }
+
+    let mut in_i = 0usize;
+    let mut remaining = compressed.len();
+
+    let mut out = vec![0u8; expected_byte_size];
+    let mut tmp_u16_buffer = Vec::<u16>::with_capacity(expected_byte_size / 2);
+    let mut out_i = 0;
+
+    debug_assert_eq!(tmp_read_index, expected_value_count);
+
+    for channel in &mut channel_data {
+        // Sample types that does not support B44 compression (u32 and f32) are raw copied.
+        if channel.type_ != SampleType::F16 {
+            // Compute byte count for current channel.
+            let byte_count = channel.resolution.area()
+                * channel.samples_per_pixel
+                * channel.type_.bytes_per_sample();
+
+            if remaining < byte_count {
+                // notEnoughData();
+                return Err(Error::invalid("not enough data"));
+            }
+
+            //memcpy (cd.start, inPtr, n);
+            for i in channel.tmp_start_index..(channel.tmp_start_index + byte_count) {
+                // TOTALLY WRONG
+                out[out_i + i] = compressed[in_i + i];
+            }
```

Comment:
I need to review this part. It came from a moment when I didn't use a `tmp_u16_buffer`.

Reply:
`memcpy(dest_ptr, src_ptr, count)` can most of the time be replaced with `dest_slice.copy_from_slice(src_slice)`, if you have two slices of the same length. You can create those slices using `let src_slice = &source_slice[start_index .. end_index]`. 

Your code is already doing this. If this is wrong, than using slices might not solve the issue though.